### ISSUE TITLE
Argument order of `String.split`, `String.splitAt` and `String.splitOn`???

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#2257]
First, why do these functions take their "subject" first? Is this because back when we thought UFCS was a solution to all our problems we wanted that?

Second, `split` and `splitOn` takes two strings. I find this really dangerous (easy to mix up). What can be done?

**Comment by mlutze** [comment]
For `split` we can take a proper `Regex` type rather than a string. I'm not sure what the solution for `splitOn` is.

**Comment by magnus-madsen** [comment]
@jaschdoc Could I ask you to compile a list of String operations that are *not* subject-last? 

Subject-last means that if you have a call `f(..., "the string you are working on must be the last argument"). As you can see `split` and `splitOn` violates this and should be in the list.

**Comment by jaschdoc** [comment]
# List of Not Subject-last String Operations

1. `slice`
2. `sliceLeft`
3. `sliceRight`
4. `repeat`
5. `stripIndentDropWhiteSpace` (this is a private helper function)
6. `contains`
7. `splitOn`

**Comment by magnus-madsen** [comment]
@ninaandrup 

I would like to have a PR that attempts the following:

- [ ] Changes the signature of `String.contains` to:

```
pub def contains({needle: String}, haystack: String): Bool
```

- [ ] Changes the signature of `String.startsWith` to:

```
pub def startsWith({prefix: String}, s: String): Bool
```

Here is some documentation records: https://doc.flix.dev/records/

**Comment by mlutze** [comment]
Instead of `needle`, would it be better to have a more literal term like `substring`?

**Comment by magnus-madsen** [comment]
I would to have these String signatures updated:

```scala
pub def endsWith(suffix: {suffix: String}, s2: String): Bool
pub def contains(substring: {substring: String}, s: String): Bool
pub def splitOn(substring: {substring}, s: String): List[String]
pub def split(regex: {regex: String}, s: String): List[String] // Note: Updated return type and effect.
pub def isMatch(regex: {regex: String}, s: String)
pub def isSubmatch(regex: {regex: String}, s: String)
pub def indices(substring: {substring: String}, s: String)
pub def replace(target: {target: String}, replacement: {replacement: String}, s: String): String
```

Note: There will be a few more to come. I will update the list later.

based on @mlutze comment, I would also like:

```
pub def contains(n: {needle: String}, haystack: String): Bool
```

changed to

```
pub def contains(substring: {substring: String}, s: String): Bool
```

`levenshteinDistance` is symmetric and so should NOT be changed.
`stripIndentDropWhiteSpace ` is a helper and will be left as is.
`repeat`, `slice` etc. are not relevant.

**Comment by mlutze** [comment]
> We are not worried that using an abbreviation is a problem?

Is your worry that `substr` wouldn't be understandable? Or that it would be hard to remember?

**Comment by magnus-madsen** [comment]
Fixed in #2421.
